### ISSUE TITLE
Infinite recursion on findOneAndUpdate with nested docs

### ISSUE THREAD

**Issue by dojadop** [Automattic/mongoose#4477]
v4.5.10

This commit breaks using findOneAndUpdate https://github.com/Automattic/mongoose/commit/b15b5b6f2e965c9a0b9310c437b45b52a149b53b with nested docs

```
  modelWithNestedDoc.findOneAndUpdate({key: json.key}, json, {upsert: true}).then((result) => {
...
  });
```

This function setDefaultsOnInsert in query.js (https://github.com/Automattic/mongoose/blob/master/lib/query.js#L1826) eventually calls https://github.com/Automattic/mongoose/blob/master/lib/services/common.js#L53

This function will infinite loop on nested documents since they are casted to Document objects instead of an Object.

v4.5.9: nested objects are of type Object
v4.5.10: nested objects are of type Document

When it is a Document the modifiedPaths infinitely recurses on the nested properties of the Document object trying to flatten it out.

```
$__ = InternalCache
$__original_remove = function () {
$__original_save = function () {
 arguments = null
 caller = null
 length = 0
 name = ""
 numAsyncPres = 1
...
```

**Comment by davebaol** [comment]
I got  a similar issue (infinite recursion) after updating to 4.5.10.
4.5.8 works good.

**Comment by vkarpov15** [comment]
Duplicate of #4465 should be fixed when 4.6.1 is released
